File: src/trajectory.cpp
#include "../header/trajectory.h"

#include <iostream>
#include <fstream>
#include <vector>
#include <filesystem>
#include <algorithm>
#include <thread>

#include "arrow/api.h"
#include "arrow/io/api.h"
#include <parquet/arrow/writer.h>

// Constructor
Trajectory::Trajectory():R(6371000), altInterval(50), tangentFactor(10){
    trueDistance = 0;
}

float Trajectory::evaluateSigmoid(float x){

    float sigmoid_factor = 1/(1 + exp(-(x-0.5*alt)/(alt/tangentFactor)));
    return downrangeDistance * sigmoid_factor;

}

void Trajectory::calculateTrajectory(){
    int temp_count = 0;
    
    for(int i = 0; i <= (int)alt; i += altInterval){
        temp_count++;

        xTrajectoryValues.push_back(evaluateSigmoid((float)i));
        yTrajectoryValues.push_back(-(float)i + alt);

        if(temp_count >= 2){
            trueDistance += sqrt(pow(xTrajectoryValues[temp_count-1]-xTrajectoryValues[temp_count-2], 2) + pow(yTrajectoryValues[temp_count-1]-yTrajectoryValues[temp_count-2], 2));
        }
    }

    if(yTrajectoryValues.back() != 0){
        xTrajectoryValues.push_back(evaluateSigmoid((float)alt));
        yTrajectoryValues.push_back(0);
    }

    numberOfPoints = xTrajectoryValues.size();

    // Store data
    // Generate File
    std::string run_data_path = generateOutputPath();
    std::ofstream output_csv(run_data_path);
    if (!output_csv.is_open()) {
        std::cerr << "Error generating file." << std::endl;
        return;
    }

    // Store to run-specific csv
    arrow::Status returnStatus = recordParquetLog(run_data_path, xTrajectoryValues, yTrajectoryValues, vehicleSpeed, thrust);
    if (!returnStatus.ok()) {
        std::cerr << "Failed to record Parquet log: " << returnStatus.ToString() << std::endl;
        return;
    }
    

    // Store to manifest
    float lls[2] = {latStart, lonStart};
    float lle[2] = {latEnd, lonEnd};
    logToManifest(run_data_path, lls, lle, alt, isValidTrajectory());

    output_csv.close(); 

}

void Trajectory::resetTrajectory(){
    xTrajectoryValues.clear();
    yTrajectoryValues.clear();
    numberOfPoints = trueDistance = alt = downrangeDistance = 0;
}

bool Trajectory::isValidTrajectory(){
    float phi1 = latStart * (M_PI / 180.0);
    float phi2 = latEnd * (M_PI / 180.0);
    float lam1 = lonStart * (M_PI / 180.0);
    float lam2 = lonEnd * (M_PI / 180.0);
    float dphi = phi2 - phi1;
    float dlam = lam2 - lam1;

    float a = pow(sin(dphi / 2), 2) + cos(phi1) * cos(phi2) * pow(sin(dlam / 2), 2);
    downrangeDistance = R * atan2(sqrt(a), sqrt(1 - a));
    // downrange_distance = 2 * R * asin(sqrt(a));

    lineOfSightAngle = atan(alt / downrangeDistance) * (180 / M_PI);

    bearing = (atan2(sin(dlam) * cos(phi2), cos(phi1) * sin(phi2) - sin(phi1) * cos(phi2) * cos(dlam))) * (180 / M_PI);

    minDistance = sqrt(pow(alt, 2) + pow(downrangeDistance, 2));

    if (lineOfSightAngle < 0.0) {
        return false;
    }

    return true;
}

// -----

namespace fs = std::filesystem;

std::string Trajectory::generateOutputPath() {
    // Locate data folder
    fs::path target_path = fs::canonical(fs::read_symlink("/proc/self/exe"));
    target_path = target_path.parent_path().parent_path().append("data");

    // Count number of files
    int count;
    try {
        count = std::count_if(
            fs::directory_iterator(target_path),
            fs::directory_iterator(),
            [](const auto& entry) {
                return entry.is_regular_file();
            }
        );
    } catch (const fs::filesystem_error& e) {
        std::cerr << "Error: " << e.what() << '\n';
        return 0;
    }

    // Create file name
    std::stringstream file;
    file << std::to_string(count + 1) << "-" << std::this_thread::get_id() << ".parquet";
    std::string temp_file = file.str();

    // Return full file path
    return target_path.append(temp_file);
    
}

bool Trajectory::logToManifest(const std::string& output_csv,
                               const float* lat_lon_start,
                               const float* lat_lon_end,
                               const float& altitude,
                               const bool& feasible){
    // Locate manifest
    fs::path target_path = fs::canonical(fs::read_symlink("/proc/self/exe"));
    target_path = target_path.parent_path().parent_path().append("data/manifest.csv");

    try{
        std::ofstream manifest_csv(target_path, std::ios::app);

        // Store to manifest
        manifest_csv 
        << lat_lon_start[0] 
        << ',' 
        << lat_lon_start[1] 
        << ',' 
        << lat_lon_end[0]
        << ','
        << lat_lon_end[1]
        << ','
        << altitude
        << ','
        << feasible
        << ','
        << output_csv 
        << std::endl; 

        manifest_csv.close();
        return 0;
    } catch (const std::runtime_error& error) {
        std::cerr << "Error: " << error.what() << std::endl;
        return 1; 
    }

}

// bool Trajectory::recordParquetLog(std::string output_csv, std::vector<float> downrange_dist, std::vector<float> altitude, std::vector<float> vehicle_speed, std::vector<float> thrust){        

//     try {
//         std::ofstream output(output_csv);

//         // Write column titles
//         output << "downrange_dist, alititude, vehicle_speed, thrust" << std::endl;

//         for(int i = 0; i < downrange_dist.size(); i++){
//             output << downrange_dist[i] << ',' << altitude[i] << ',' << "temp" << ',' << "temp" << std::endl;
//         }

//         output.close();
//         return 0;
//     } catch (const std::runtime_error& error){
//         std::cerr << "Error: " << error.what() << std::endl;
//         return 1; 
//     }

// }
    
// bool Trajectory::recordParquetLog(const std::string& output_path,
//                                   const std::vector<float>& downrange_dist,
//                                   const std::vector<float>& altitude,
//                                   const std::vector<float>& vehicle_speed,
//                                   const std::vector<float>& thrust) {
//     try {
//         arrow::FloatBuilder downrange_builder;
//         arrow::FloatBuilder altitude_builder;
//         arrow::FloatBuilder speed_builder;
//         arrow::FloatBuilder thrust_builder;

//         // Build Arrow arrays
//         for (size_t i = 0; i < downrange_dist.size(); ++i) {
//             downrange_builder.Append(downrange_dist[i]);
//             altitude_builder.Append(altitude[i]);
//             speed_builder.Append(vehicle_speed[i]);
//             thrust_builder.Append(thrust[i]);
//         }

//         std::shared_ptr<arrow::Array> downrange_array;
//         std::shared_ptr<arrow::Array> altitude_array;
//         std::shared_ptr<arrow::Array> speed_array;
//         std::shared_ptr<arrow::Array> thrust_array;

//         downrange_builder.Finish(&downrange_array);
//         altitude_builder.Finish(&altitude_array);
//         speed_builder.Finish(&speed_array);
//         thrust_builder.Finish(&thrust_array);

//         auto schema = arrow::schema({
//             arrow::field("downrange_dist", arrow::float32()),
//             arrow::field("altitude", arrow::float32()),
//             arrow::field("vehicle_speed", arrow::float32()),
//             arrow::field("thrust", arrow::float32())
//         });

//         auto table = arrow::Table::Make(schema, {
//             downrange_array, altitude_array, speed_array, thrust_array
//         });

//         std::shared_ptr<arrow::io::FileOutputStream> outfile;
//         ARROW_ASSIGN_OR_RAISE(outfile, arrow::io::FileOutputStream::Open(output_path));

//         PARQUET_THROW_NOT_OK(parquet::arrow::WriteTable(
//             *table, arrow::default_memory_pool(), outfile, 1024
//         ));

//         return true;
//     } catch (const std::exception& e) {
//         std::cerr << "Error writing Parquet: " << e.what() << std::endl;
//         return false;
//     }
// }

arrow::Status Trajectory::recordParquetLog(
    const std::string& output_path,
    const std::vector<float>& downrange_dist,
    const std::vector<float>& altitude,
    const std::vector<float>& vehicle_speed,
    const std::vector<float>& thrust)
{
    arrow::FloatBuilder downrange_builder;
    arrow::FloatBuilder altitude_builder;
    arrow::FloatBuilder speed_builder;
    arrow::FloatBuilder thrust_builder;

    for (size_t i = 0; i < downrange_dist.size(); ++i) {
        ARROW_RETURN_NOT_OK(downrange_builder.Append(downrange_dist[i]));
        ARROW_RETURN_NOT_OK(altitude_builder.Append(altitude[i]));
        ARROW_RETURN_NOT_OK(speed_builder.Append(vehicle_speed[i]));
        ARROW_RETURN_NOT_OK(thrust_builder.Append(thrust[i]));
    }

    std::shared_ptr<arrow::Array> downrange_array;
    std::shared_ptr<arrow::Array> altitude_array;
    std::shared_ptr<arrow::Array> speed_array;
    std::shared_ptr<arrow::Array> thrust_array;

    ARROW_RETURN_NOT_OK(downrange_builder.Finish(&downrange_array));
    ARROW_RETURN_NOT_OK(altitude_builder.Finish(&altitude_array));
    ARROW_RETURN_NOT_OK(speed_builder.Finish(&speed_array));
    ARROW_RETURN_NOT_OK(thrust_builder.Finish(&thrust_array));

    auto schema = arrow::schema({
        arrow::field("downrange_dist", arrow::float32()),
        arrow::field("altitude", arrow::float32()),
        arrow::field("vehicle_speed", arrow::float32()),
        arrow::field("thrust", arrow::float32())
    });

    auto table = arrow::Table::Make(schema, {
        downrange_array,
        altitude_array,
        speed_array,
        thrust_array
    });

    ARROW_ASSIGN_OR_RAISE(auto outfile,
        arrow::io::FileOutputStream::Open(output_path));

    PARQUET_THROW_NOT_OK(parquet::arrow::WriteTable(
        *table, arrow::default_memory_pool(), outfile, 1024));

    return arrow::Status::OK();
}

File: python implementation/trajectory.py
import os
import math
import threading
import inspect
from pathlib import Path

import pyarrow as pa
import pyarrow.parquet as pq


class Trajectory:
    def __init__(self, alt_interval: int = 50, tangent_factor: int = 10, R: float = 6_371_000):
        # Earth radius, sampling interval, and curve factor
        self.R = R
        self.alt_interval = alt_interval
        self.tangent_factor = tangent_factor

        # Trajectory state
        self.true_distance = 0.0
        self.downrange_distance = 0.0
        self.line_of_sight_angle = 0.0
        self.bearing = 0.0
        self.min_distance = 0.0
        self.x_trajectory_values: list[float] = []
        self.y_trajectory_values: list[float] = []
        self.number_of_points = 0

        # User-defined parameters (set before calling calculate_trajectory)
        self.alt = 0.0  # peak altitude
        self.vehicle_speed = 0.0
        self.thrust = 0.0
        self.lat_start = 0.0
        self.lon_start = 0.0
        self.lat_end = 0.0
        self.lon_end = 0.0

    def evaluate_sigmoid(self, x: float) -> float:
        """
        Compute the downrange distance scaled by a sigmoid curve of altitude.
        """
        factor = 1 / (1 + math.exp(-(x - 0.5 * self.alt) / (self.alt / self.tangent_factor)))
        return self.downrange_distance * factor

    def calculate_trajectory(self) -> None:
        """
        Sample the trajectory curve, compute true distance, write a Parquet log,
        and append an entry to the manifest CSV.
        """
        # Reset
        self.x_trajectory_values.clear()
        self.y_trajectory_values.clear()
        self.true_distance = 0.0

        # Sample down-and-back curve
        count = 0
        for h in range(0, int(self.alt) + 1, self.alt_interval):
            count += 1
            x_val = self.evaluate_sigmoid(float(h))
            y_val = -float(h) + self.alt
            self.x_trajectory_values.append(x_val)
            self.y_trajectory_values.append(y_val)
            if count >= 2:
                dx = self.x_trajectory_values[-1] - self.x_trajectory_values[-2]
                dy = self.y_trajectory_values[-1] - self.y_trajectory_values[-2]
                self.true_distance += math.hypot(dx, dy)

        # Ensure it ends at ground level
        if self.y_trajectory_values[-1] != 0.0:
            x_val = self.evaluate_sigmoid(self.alt)
            self.x_trajectory_values.append(x_val)
            self.y_trajectory_values.append(0.0)

        self.number_of_points = len(self.x_trajectory_values)

        # Write data files
        out_path = self._generate_output_path()
        self._record_parquet_log(out_path)
        self._log_to_manifest(out_path)

    def reset_trajectory(self) -> None:
        """Clear all computed and parameter values."""
        self.__init__(self.alt_interval, self.tangent_factor, self.R)

    def is_valid_trajectory(self) -> bool:
        """
        Validate based on line-of-sight angle and compute geometrical metrics.
        """
        phi1 = math.radians(self.lat_start)
        phi2 = math.radians(self.lat_end)
        dphi = phi2 - phi1
        dlam = math.radians(self.lon_end - self.lon_start)

        a = (math.sin(dphi / 2) ** 2
             + math.cos(phi1) * math.cos(phi2) * math.sin(dlam / 2) ** 2)
        self.downrange_distance = self.R * math.atan2(math.sqrt(a), math.sqrt(1 - a))
        self.line_of_sight_angle = math.degrees(math.atan(self.alt / self.downrange_distance))
        self.bearing = math.degrees(math.atan2(
            math.sin(dlam) * math.cos(phi2),
            math.cos(phi1) * math.sin(phi2)
            - math.sin(phi1) * math.cos(phi2) * math.cos(dlam)
        ))
        self.min_distance = math.hypot(self.alt, self.downrange_distance)

        return self.line_of_sight_angle >= 0.0

    def _generate_output_path(self) -> str:
        """
        Create a unique parquet filename in the ../data directory next to this script.
        """
        # Locate data folder two levels up
        script_file = inspect.getsourcefile(lambda: None)
        base_dir = Path(script_file).resolve().parent.parent / "data"
        base_dir.mkdir(parents=True, exist_ok=True)

        # Count existing files
        existing = [f for f in base_dir.iterdir() if f.is_file()]
        filename = f"{len(existing) + 1}-{threading.get_ident()}.parquet"
        return str(base_dir / filename)

    def _record_parquet_log(self, path: str) -> None:
        """
        Write trajectory samples and constants to a Parquet file.
        """
        n = len(self.x_trajectory_values)
        table = pa.Table.from_pydict({
            "x": self.x_trajectory_values,
            "y": self.y_trajectory_values,
            "vehicle_speed": [self.vehicle_speed] * n,
            "thrust": [self.thrust] * n
        })
        pq.write_table(table, path)

    def _log_to_manifest(self, path: str) -> None:
        """
        Append a CSV entry summarizing the run parameters and file path.
        """
        script_file = inspect.getsourcefile(lambda: None)
        manifest = Path(script_file).resolve().parent.parent / "data" / "manifest.csv"
        manifest.parent.mkdir(exist_ok=True)

        feasible = self.is_valid_trajectory()
        line = (
            f"{self.lat_start},{self.lon_start},"
            f"{self.lat_end},{self.lon_end},"
            f"{self.alt},{feasible},{path}\n"
        )
        with open(manifest, "a", newline="") as f:
            f.write(line)
